kaseyjsmith/chart builder revamp (#313)
embrace-io/embrace-docs: docs/product/custom-dashboards.md

<img src={require('@site/static/images/custom_dashboards/Add_Widget_Menu.png').default} style={{ width: '75%', height: '75%' }} alt="Create new widget" />

## Creating a Chart
You can create multiple types of visualizations. A table of which visualizations are supported is shown below:

| Chart Type | Crashes | Network | Logs | Issues | Moments | Sessions | Spans | Historical | Multi-query |
| ---------- | ------- | ------- | ---- | ------ | ------- | -------- | ----- | ---------- | ----------- |
| Line | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |
| Table | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x: | :x:
| Bar | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x: | :x: |
| Pie | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x: | :x: |

## Spans (Traces)

> Create graphs monitoring Spans performance. You can filter and group by name, outcome, duration, and any attributes you have set on the Span.

Author: ```suggestion
Create graphs monitoring Spans performance. You can filter and group by name, outcome, duration, and any custom Attributes you have set on the Span.
```

- First, select the visualization type you want to see and a source for the query.
- Then, choose the metric you wish to aggregate. For Spans, Embrace supports both counts and sums of duration.
- Finally, add any filters and group-bys. In addition to our core dimensions, for Spans you can use the Span's name, outcome, duration, and any custom Attributes.

<img src={require('@site/static/images/custom_dashboards/Spans_Chart_Builder.png').default} style={{ width: '75%', height: '75%' }} alt="Spans as a Widget option" />

## Combined Time Series (Line Charts)

You can also create Combined graphs in Line vizualizations. With Combined graphs, you can visualize up to 10 time series together and optionally create a custom time series using arithmetic operators.

**NOTE:** as of 2025-05-20, the chart builder has been improved. Any legacy charts using fills have been moved to multi-query line charts. 

### Adding Formulas to Graphs

> For example, you want to get a rate over the last 24 hours of the sum of crashes and ANR's with a duration of 3.5 to 10 second relatve to your total sessions, grouped by App Version. To do this, you will:

Reviewer: ```suggestion
For example, you want to get a rate over the last 24 hours of the sum of crashes and ANR's with a duration of 3.5 to 10 second relatve to your total sessions, grouped by App Version. To do this:
```

The way this reads to me is `To do this, you will for the initial query A` which sounds odd.

> 1. For the inital query A, select Crashes as the source, Crash Count as the metric, and App Version in the group by field.\

Reviewer: There's some inconsistency with `Query` vs `query` here.

    a. You can also add a group by here. For this example, we group by App Version. See more on grouping functionality below.\
    b. Formulas and multiple queries are only supported on line charts currently. 

> 2. On Query A, click the Filter By button below the query details and filter for State = foreground. 

Author: ```suggestion
2. On query A, click the Filter By button below the query details and filter for State = foreground. 
```

3. Create a second query (B) by clicking the "Add query" button. Select ANR's as the source, ANR count as the metric, and again group by App Version.\

>     a. On Query B, filter for Duration Bucket (milliseconds) = 3500 - 10000.

Author: ```suggestion
    a. On query B, filter for Duration Bucket (milliseconds) = 3500 - 10000.
```

4. Create a third query (C) for your Session Count grouped by App Version and filter for foreground sessions using the same process as in query A.
5. Hide queries A, B, and C by clicking the eye symbols at the right side of each query row.
5. In the Formula field at the bottom, enter `(A+B) / C`. You can optionally multiply this by 100 to get a y-axis as a percentage (`((A+B) / C) * 100`).\
    a. The Formula field accepts the operators: `+, -, *, /`.\
    b. Order of operations follows standard PEMDAS.

<img src={require('@site/static/images/custom_dashboards/Formula_Timeseries_Example.png').default} style={{ width: '75%', height: '75%' }} alt="Multiple queries and formula" />

### Multiple Combined Time Series without a Formula

> Since the Formula field is optional, you can visualize up to 10 time series on the same chart with the steps above. Use the eye symbols to show or hide each and leave the Formula field blank.

Reviewer: I know this is already in the docs

```suggestion
Since the Formula field is optional, you can visualize up to 10 time series on the same chart with the steps above. Use the eye symbols to show or hide each, and leave the Formula field blank.
```

| `{"App Version": "2.0"}` | 50 | 10 | 60 |
 
- Grouping on both time series

    - `A`: *grouped by* `[”App Version”, “Country”]`
        - Returns values for each distinct, non-empty combination of grouping dimensions. 
    - `B`: *grouped by* `[”App Version”]`
        - Returns values for each distinct, non-empty `App Version`.
    - Both `A` and `B` have groupings, and `B` is a subset of `A`. Their common subset is `[”App Version”]`.

| Dimension | A | B | A + B | 
| :--------------: | :---------------: | :---------------: |:---------------: |
| `{"App Version": "1.0", "Country": "US"}` | 100 | 10 | 110 | 
| `{"App Version": "1.0", "Country": "MEX"}` | 30 | 10 | 40 |
| `{"App Version": "2.0", "Country": "US"}` | 50 | 20 | 70 |
| `{"App Version": "2.0", "Country": "MEX"}` | 60 | 20 | 80 |

- No groupings

    - `A`: *grouped by* `[]` 
    - `B`: *grouped by* `[]` 
    - `A` and `B` have no groupings. Their common join key is the empty set.
 
#### **Invalid Groupings:**

- Grouping with no overlapping set

    - `A`: *grouped by* `[”App Version”]` 
    - `B`: *grouped by* `[”OS Version”]` 
    - Both have groupings but neither is a subset of the other.

> - Grouping 

Reviewer: Is this description unfinished?

> - Grouping 

Author: ```suggestion
- Grouping with similar Attributes, but no overlapping data
```

    - `A`: *grouped by* `[”App Version”, “Country”, “OS Version”]`
    - `B`: *grouped by* `[”App Version”, “Country”, “Model”]`
    - If A and B have no similar `App Version` and `Country`, the grouping is valid, but neither is a subset of the other and no results are returned.

#### Order of Operations

Additionally, order of operations matters. Consider a situation where you create three queries:

- `A` grouped by `["App Version", "OS Version"]` 
- `B` grouped by `["App Version"]` 
- `C` grouped by `["OS Version"]` 

Then you were to combine them in the formula field:

- **Valid:** `(A + B) + C`: `A` and `B` share a grouping dimension and can combine, then combine with `C`.

> - **Invalid:** `A + (B + C)`: `B` and `C` share now group-by dimension and therefore fail. 

Reviewer: Is this supposed to be `share no group-by dimension`? Maybe `do not share a dimension`?

> - **Invalid:** `A + (B + C)`: `B` and `C` share now group-by dimension and therefore fail. 

Author: ```suggestion
- **Invalid:** `A + (B + C)`: `B` and `C` do not share a group-by dimension and therefore fail. 
```
